### ISSUE TITLE
Does it support CTM training from scratch?

### ISSUE THREAD

**Issue by Keely-Ai** [Kim-Dongjun/ctm-cifar10#3]
Hi! I am wondering if I set self_learn to true, is it the training from scratch version?
Thank you so much!

**Comment by Kim-Dongjun** [comment]
Yes, this code is an original one that releases all things. All you need to do is to set up the hyperparameters properly

**Comment by Keely-Ai** [comment]
> Yes, this code is an original one that releases all things. All you need to do is to set up the hyperparameters properly

Thank you very much for the quick response!
I'm very interested in CTM training from scratch and one-step sampling. I’d like to confirm whether we should set `num_heun_step` = 17 or `num_heun_step` = 1. Also, do we need to set `self_learn_iterative` to `True` in order to achieve one-step CTM results?
I really appreciate your help!

**Comment by yuanzhi-zhu** [comment]
@Kim-Dongjun Could you please provide the command line for training from scratch?
It would be even better if you still have the ckpt of ctm trained from scratch (cifar10 uncond), or you still have the fid of it.

**Comment by Kim-Dongjun** [comment]
Hi all! Thanks for your interest. Here I give you some command lines that I used before.

mpiexec -n 4 python3.11 cm_train.py --out_dir /groups/gce50978/user/dongjun/EighthArticleExperimentalResults/CIFAR10/self_learn/CTM_default/ --microbatch=16 --global_batch_size=128 --self_learn=True --eval_interval=10000 --save_interval=10000

(or some other command that I used was "mpiexec -n 4 python3.11 cm_train.py $MODEL_FLAGS --out_dir /home/acf15618av/EighthArticleExperimentalResults/CIFAR10/self_learn/CTM/CTM_self_M=1_dsm_no_adaptive/ --num_heun_step=1 --num_heun_step_random=False" Try this out if the above one does not work)

and then after some iterations you can include GAN training as below:

mpiexec -n 4 python3.11 cm_train.py --out_dir /home/acf15618av/EighthArticleExperimentalResults/CIFAR10/self_learn/GAN_default/ --gan_fake_inner_type=model --gan_fake_outer_type=target_model_sg --gan_training=True --g_learning_period=2 --resume_checkpoint=/home/acf15618av/EighthArticleExperimentalResults/CIFAR10/resume_checkpoint/euler_edm_stop_grad_random_17_M_bs_384_from_edm_dsm_1.0/model050000.pt --microbatch=10 --global_batch_size=120 --self_learn=True

For your information, I attach you the log files and training configs as below.

[GAN_finetuning_training_config_0.txt](https://github.com/user-attachments/files/19798610/GAN_finetuning_training_config_0.txt)
[GAN_finetuning_log.txt](https://github.com/user-attachments/files/19798612/GAN_finetuning_log.txt)
[CTM_pretraining_log.txt](https://github.com/user-attachments/files/19798611/CTM_pretraining_log.txt)
[CTM_pretraining_training_config_0.txt](https://github.com/user-attachments/files/19798609/CTM_pretraining_training_config_0.txt)

**Comment by Kim-Dongjun** [comment]
From my personal experience, training CTM from scratch tends to be a bit unstable or unpleasant, but honestly, it wasn’t that bad. So I think there’s definitely room to write a paper about it. If you're interested in this topic, it might be a good idea to highlight how unstable CTM training from scratch can be, and then explore some stabilization techniques. I think that could lead to a solid research direction.

**Comment by Keely-Ai** [comment]
> Hi all! Thanks for your interest. Here I give you some command lines that I used before.
> 
> mpiexec -n 4 python3.11 cm_train.py --out_dir /groups/gce50978/user/dongjun/EighthArticleExperimentalResults/CIFAR10/self_learn/CTM_default/ --microbatch=16 --global_batch_size=128 --self_learn=True --eval_interval=10000 --save_interval=10000
> 
> (or some other command that I used was "mpiexec -n 4 python3.11 cm_train.py $MODEL_FLAGS --out_dir /home/acf15618av/EighthArticleExperimentalResults/CIFAR10/self_learn/CTM/CTM_self_M=1_dsm_no_adaptive/ --num_heun_step=1 --num_heun_step_random=False" Try this out if the above one does not work)
> 
> and then after some iterations you can include GAN training as below:
> 
> mpiexec -n 4 python3.11 cm_train.py --out_dir /home/acf15618av/EighthArticleExperimentalResults/CIFAR10/self_learn/GAN_default/ --gan_fake_inner_type=model --gan_fake_outer_type=target_model_sg --gan_training=True --g_learning_period=2 --resume_checkpoint=/home/acf15618av/EighthArticleExperimentalResults/CIFAR10/resume_checkpoint/euler_edm_stop_grad_random_17_M_bs_384_from_edm_dsm_1.0/model050000.pt --microbatch=10 --global_batch_size=120 --self_learn=True
> 
> For your information, I attach you the log files and training configs as below.
> 
> [GAN_finetuning_training_config_0.txt](https://github.com/user-attachments/files/19798610/GAN_finetuning_training_config_0.txt) [GAN_finetuning_log.txt](https://github.com/user-attachments/files/19798612/GAN_finetuning_log.txt) [CTM_pretraining_log.txt](https://github.com/user-attachments/files/19798611/CTM_pretraining_log.txt) [CTM_pretraining_training_config_0.txt](https://github.com/user-attachments/files/19798609/CTM_pretraining_training_config_0.txt)

Thank you so much for the quick and detailed response! The config really helps a lot, I will try that.

**Comment by yuanzhi-zhu** [comment]
@Kim-Dongjun I tried to finetune with teacher from step 0, this is the result I got after 20k steps:

> SONY-2025-04-18-13-16-00
> 19999-th step exact sampler (NFE 1) seed 42 EMA 0.0 FID-50k: 7.655472892120612
> 19999-th step exact sampler (NFE 2) seed 42 EMA 0.0 FID-50k: 6.481291077644987
> Evaluation with model parameter end
> 19999-th step exact sampler (NFE 1) seed 42 EMA 0.999 FID-50k: 7.113211170076509
> 19999-th step exact sampler (NFE 2) seed 42 EMA 0.999 FID-50k: 4.929306404602029

Do you have the log for this so I can compare?
